src/game.cpp

```diff
@@ -13113,33 +13127,159 @@ int game::slip_down_chance( bool show_messages )
         }
     }
 
+    // Affordances (other than ledge) may reduce slip chance, even below zero.
+    slip += aid.slip_chance_mod;
+    if( show_chance_messages ) {
+        if( aid.slip_chance_mod >= 0 ) {
+            // TODO allow for a message specific to the climbing aid?
+            add_msg( m_info, _( "There's nothing here to help you climb." ) );
+        }
+    }
+
+    add_msg_debug( debugmode::DF_GAME, "After affordance modifier, final slip chance %d%%",
+                   slip );
+
     return slip;
 }
 
-bool game::slip_down( bool check_for_traps, bool show_chance_messages )
+bool game::slip_down( climb_maneuver maneuver, climbing_aid_id aid_id,
+                      bool show_chance_messages )
 {
-    int slip = slip_down_chance( show_chance_messages );
+    int slip = slip_down_chance( maneuver, aid_id, show_chance_messages );
 
     if( x_in_y( slip, 100 ) ) {
         add_msg( m_bad, _( "You slip while climbing and fall down." ) );
         if( slip >= 100 ) {
             add_msg( m_bad, _( "Climbing is impossible in your current state." ) );
         }
-        if( check_for_traps ) {
+        // Check for traps if climbing UP or DOWN.  Note that ledges (open air) count as traps.
+        if( maneuver != climb_maneuver::over_obstacle ) {
             m.creature_on_trap( u );
         }
         return true;
     }
     return false;
 }
 
-void game::climb_down( const tripoint &examp )
+static int climb_affordance_menu_encode( const climbing_aid_id &aid_id )
+{
+    return 0x1000 + int_id<climbing_aid>( aid_id ).to_i();
+}
+static bool climb_affordance_menu_decode( int retval, climbing_aid_id &aid_id )
+{
+    int_id< climbing_aid > as_int_id( retval - 0x1000 );
+    if( as_int_id.is_valid() ) {
+        aid_id = as_int_id.id();
+        return true;
+    }
+    return false;
+}
```

Comment:
A comment explaining the purpose of these functions would be helpful, they're a little opaque at first glance.

Reply:
Good point.  Resolved in forthcoming commit to #68265.